neuraxle/base.py

```diff
@@ -3621,6 +3648,108 @@ def load(self, context: ExecutionContext, full_dump=True) -> BaseStep:
         return loaded_self.load(context, full_dump)
 
 
+def assert_has_services(has_service_assertions: List[Type], context: ExecutionContext):
+    """
+    Assert that all the necessary services are provided in the execution context.
+
+    :param context: The ExecutionContext for which we test the presence of service.
+    """
+    for has_service_assertion in has_service_assertions:
+        if not context.has_service(service_abstract_class_type=has_service_assertion):
+            exception_message: str = '{} dependency missing in the ExecutionContext. Please register the service {} inside the ExecutionContext.\n'.format(
+                has_service_assertion.__name__,
+                has_service_assertion.__name__
+            )
+            step_method_message: str = 'You can do so by calling register_service, or set_services on any step.\n'
+            execution_context_methods_messsage: str = 'There is also the option to register all services inside the ExecutionContext'
+            raise AssertionError(exception_message + step_method_message + execution_context_methods_messsage)
+
+
+class AssertionMixin:
+    @abstractmethod
+    def _assert(self, data_container: DataContainer, context: ExecutionContext):
+        pass
+
+    def _will_process(self, data_container: DataContainer, context: ExecutionContext) -> (
+            DataContainer, ExecutionContext):
+        """
+        Calls self._assert(data_container,context)
+        """
+        data_container, context = super()._will_process(data_container, context)
+        self._assert(data_container, context)
+
+        return data_container, context
+
+
+class AssertExpectedOutputNullMixin(AssertionMixin):
```

Comment:
My code review stops here. TODO note to self: continue code review.

Reply:
Not much left :P